server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/DataLineJamesMessageHookHandler.java

```diff
@@ -137,61 +142,33 @@ public Response onLine(SMTPSession session, ByteBuffer lineByteBuffer, LineHandl
                     .orElseThrow(() -> new RuntimeException("'" + SMTPConstants.DATA_MIMEMESSAGE_STREAMSOURCE.asString() + "' has not been filled."));
                 OutputStream out;
                 out = mmiss.getWritableOutputStream();
-                for (MessageHook rawHandler : mHandlers) {
-                    LOGGER.debug("executing james message handler {}", rawHandler);
-                    long start = System.currentTimeMillis();
-
-                    HookResult hRes = rawHandler.onMessage(session, new MailToMailEnvelopeWrapper(mail, out));
-                    long executionTime = System.currentTimeMillis() - start;
-
-                    if (rHooks != null) {
-                        for (HookResultHook rHook : rHooks) {
-                            LOGGER.debug("executing hook {}", rHook);
-                            hRes = rHook.onHookResult(session, hRes, executionTime, rawHandler);
-                        }
-                    }
-
-                    SMTPResponse response = AbstractHookableCmdHandler.calcDefaultSMTPResponse(hRes);
 
-                    // if the response is received, stop processing of command
-                    // handlers
-                    if (response != null) {
-                        return response;
-                    }
-                }
+                Flux<Optional<SMTPResponse>> rawHandlerResponses = Flux.fromIterable(mHandlers).map(
+                    rawHandler -> getRawHandlersSmtpResponse(session, mail, out, rawHandler));
 
-                for (JamesMessageHook messageHandler : messageHandlers) {
-                    LOGGER.debug("executing james message handler {}", messageHandler);
-                    long start = System.currentTimeMillis();
-                    HookResult hRes = messageHandler.onMessage(session, mail);
-                    long executionTime = System.currentTimeMillis() - start;
-                    if (rHooks != null) {
-                        for (HookResultHook rHook : rHooks) {
-                            LOGGER.debug("executing hook {}", rHook);
-                            hRes = rHook.onHookResult(session, hRes, executionTime, messageHandler);
-                        }
-                    }
+                Flux<Optional<SMTPResponse>> messageHandlerResponses = Flux.fromIterable(messageHandlers)
+                    .concatMap(messageHandler -> getMessageHandlersSmtpResponse(session, mail, messageHandler));
 
-                    SMTPResponse response = AbstractHookableCmdHandler.calcDefaultSMTPResponse(hRes);
+                return rawHandlerResponses
+                    .concatWith(messageHandlerResponses)
+                    // if the response is received, stop processing of command handlers
+                    .filter(Optional::isPresent)
+                    .defaultIfEmpty(Optional.empty())
+                    .subscribeOn(Schedulers.elastic())
+                    .blockFirst()
+                    .orElse(null);
```

Comment:
You do not deal with `null` results?

Would it be costly to return an `Optional`?

Reply:
I could be a refactoring, but it is outside of the scope of this PR i think.
I would like to not include new refactoring in this PR which has already grown too much.